### PR TITLE
Bug Fixing and Improvements for v2.1

### LINKED ISSUE #123: [PE-D] Invalid userinput for AddLesson
The whole input including the /code /time is added in as the lesson name!

-------------
Labels: `severity.High` `type.FunctionalityBug`
original: snowbanana12345/ped#3

### LINKED ISSUE #150: [PE-D] Too wordy to add a new lesson
I think it is a liitle bit to many words to type especially since I'm not a very fast typist

-------------
Labels: `severity.VeryLow` `type.FeatureFlaw`
original: shikai-zhou/ped#3

### LINKED ISSUE #151: [PE-D] Archive
Can you add the functionality to choose which notes to archive instead of all?

-------------
Labels: `severity.Low` `type.FeatureFlaw`
original: shikai-zhou/ped#4

### LINKED ISSUE #154: [PE-D] File parsing failure after entering module code or name containing "=-="
The saveFileTasks.txt uses the regex "=-=" to partition information. However, it can be entered in the MODULE_CODE and NAME field, resulting in incorrect parsing and loss of loading of all saved data on the subsequent runs until the issue is manually rectified.

Example with addLesson: addLesson /code CG=-=2271 /date 22-10-20 /time 17:00 /to 19:00 /repeats 55

Example with addEvent: addEvent Con=-=cert /date 13-07-19 /time 17:00 /to 21:00 /repeats 0

Example with addDeadline: addDeadline Return=-=books /by 31-12-20 23:59 /repeats 0

editLesson, editEvent and editDeadline will also result in this bug if the name/module code is modified to contain "=-=".

Edit: The same regex is used in saveFileNotes.txt and hence the same issue occurs if the user inputs a note with the title or body containing "=-=".

-------------
Labels: `severity.Medium` `type.FunctionalityBug`
original: e0426051/ped#1

### PR DESCRIPTION
**Pull request by dojh111**
Contains fixes for multiple issues and updates for UG and DG:

**Pull request summary**
* Add method to validate module codes entered. Fixes #123 
* Update UG to highlight existence of partial command format. Fixes #150 
* Update UG and DG to reflect current version and add future implementation. Fixes #151 
* Change code to defend against save delimiter being added. Fixes #154 
* Change save reading to skip over corrupted save instead of stopping process entirely.
* Update archive command to clear note history when notes archived.

### CODE REVIEW

File: src/main/java/seedu/lifeasier/parser/Parser.java [hunk 1]
```diff
@@ -67,6 +67,9 @@
 
     public static final int INDEX_START = 0;
     public static final int INDEX_END = 1;
+    public static final int MAXIMUM_CODE_LENGTH = 8;
+    public static final int MINIMUM_CODE_LENGTH = 6;
+    public static final int MAXIMUM_POSTFIX_CHARACTER = 1;
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/parser/Parser.java]: I think you can consider renaming this to suffix!
```suggestion
    public static final int MAXIMUM_SUFFIX_CHARACTER = 1;
```

File: src/main/java/seedu/lifeasier/parser/Parser.java [hunk 2]
```diff
@@ -883,6 +894,122 @@ private boolean isNumeric(String input) {
...
+
+        int letterCount = getLetterCount(moduleCodeElements);
+        //Check if module code prefix falls out of valid range of having 2-3 consecutive letters
+        if (letterCount < 2 || letterCount > 3) {
+            logger.log(Level.INFO, "Invalid prefix length");
+            return false;
+        }
+
+        int numberCount = getNumberCount(moduleCodeElements, letterCount);
+        //Check if there are 4 consecutive numbers in sequence
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/parser/Parser.java]: You could consider renaming this such that the comment is not necessary.
```suggestion
        int consecutiveNumberCount = getConsecutiveNumberCount(moduleCodeElements, letterCount);
```

File: src/main/java/seedu/lifeasier/parser/Parser.java [hunk 3]
```diff
@@ -883,6 +894,122 @@ private boolean isNumeric(String input) {
...
+        return letterCount;
+    }
+
+    /**
+     * Repeatedly prompts for valid module code until a valid one is input.
+     *
+     * @return Valid module code.
+     */
+    public String getValidModuleCode() {
+        Ui ui = new Ui();
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/parser/Parser.java]: Maybe we could find a way to do this that doesn't involve instantiating another Ui object?

Comment (daniellimzj) [reply src/main/java/seedu/lifeasier/parser/Parser.java]: I believe a ui object is already passed into parser when we call parseCommand.

image

File: src/main/java/seedu/lifeasier/storage/FileCommand.java [hunk 4]
```diff
@@ -74,4 +75,26 @@ public LocalDateTime convertToLocalDateTime(String dateTimeInformation) {
...
+     */
+    public boolean checkForDelimiterCount(String string, int baseDelimiterCount) {
+        boolean isStillContainingDelimiter = true;
+        int delimiterCount = 0;
+
+        while (isStillContainingDelimiter) {
+            isStillContainingDelimiter = string.contains(SAVE_DELIMITER);
+            if (isStillContainingDelimiter) {
+                delimiterCount++;
+                string = string.replaceFirst(SAVE_DELIMITER, BLANK_STRING);
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/storage/FileCommand.java]: Could this be refactored into a do while loop?

File: src/main/java/seedu/lifeasier/ui/Ui.java [hunk 5]
```diff
@@ -166,10 +175,28 @@ public void showHelp() {
...
+            try {
+                if (userInput.contains("=-=")) {
+                    throw new SaveDelimiterException();
+                }
+                isCommandWithoutDelimiter = true;
+            } catch (SaveDelimiterException e) {
+                System.out.println(colourTextRed("Inputs cannot contain the character sequence: =-="));
+                System.out.println(colourTextCyan("Please re-enter your input:"));
+            }
+        } while (!isCommandWithoutDelimiter);
```
Comment (daniellimzj) [inline src/main/java/seedu/lifeasier/ui/Ui.java]: Could this be simplified? It's a double negative.
```suggestion
        } while (isCommandWithDelimiter);
```

**Review by daniellimzj — COMMENTED** [review]
Looks good! Just a few small things

File: src/main/java/seedu/lifeasier/parser/Parser.java (hunk 1 above)
Comment (dojh111) [reply src/main/java/seedu/lifeasier/parser/Parser.java]: Thanks for the tip, will change

File: src/main/java/seedu/lifeasier/parser/Parser.java (hunk 2 above)
Comment (dojh111) [reply src/main/java/seedu/lifeasier/parser/Parser.java]: Good suggestion, will change the one for letter count as well.

File: src/main/java/seedu/lifeasier/storage/FileCommand.java (hunk 4 above)
Comment (dojh111) [reply src/main/java/seedu/lifeasier/storage/FileCommand.java]: Alright, noted, will change

File: src/main/java/seedu/lifeasier/ui/Ui.java (hunk 5 above)
Comment (dojh111) [reply src/main/java/seedu/lifeasier/ui/Ui.java]: Will change
